FolderPicker: Fixes issue with typing to search for folder

JoaoSilvaGrafana opened the pull request:
**What is this feature?**

It changes the debounce function used to get options for `FolderPicker` and `OpenLibraryPanelModal` as the `lodash` one does not work well with `AsyncSelect` (it only gets the "leading" search from the debounce, not the "trailing").

**Why do we need this feature?**

If a user types very quickly into the search it will return the wrong results

**Who is this feature for?**

Anyone using these two `AsyncSelect`

**Which issue(s) does this PR fix?**:

Fixes #59320

JoaoSilvaGrafana committed:
FolderPicker: Changes debounce method

JoaoSilvaGrafana committed:
Another place where it is broken

joshhunt commented:
Interesting. I was playing with FolderPicker's debouce a while back to fix some issues with it and i thought i left it on a good place... https://github.com/grafana/grafana/pull/50303

samjewell reviewed (approved):
Thanks for this! 🙏 

I tested this dev-branch locally, and it works now, where it didn't work for me before 🎉 .

Somehow when I type "j", "e" slowly, the reactions to those slow keypresses feel far more instantaneous to me. When typing super fast it wasn't _quite_ as instantaneous as I had hoped, but it still does the job and fixes the error.

I'll approve as I've tested it and it works for me.
But **I've never touched the Grafana front-end code before**, so I think it's probably worth getting a review from someone with more context than me here.

---

Given @joshhunt's comment [here](https://github.com/grafana/grafana/pull/59434#issuecomment-1329608841) I was wondering whether there should be a test for this behaviour, in order to prevent it from getting broken again. Just a question really - I don't understand the code here, or React more generally I'm afraid. But it seems to me that a test might help to properly document the behaviour we're after, and that Josh thought he had fixed, but hadn't. I'll leave it up to you - I know even less about how we'd consider testing this kind of functionality.

JoaoSilvaGrafana commented:
> Interesting. I was playing with FolderPicker's debouce a while back to fix some issues with it and i thought i left it on a good place... #50303

@joshhunt I believe there were some changes in 9.2.0 which might've broken it again: https://github.com/grafana/grafana/pull/54687/files#diff-f8e2a9b80d5f6d75c204d3c2bde8aead4d9062f8ad8df4012451dd6d50c4532bR115
They removed the `trailing` call from the debounce, which would mean only the first call would be called. But my problem was that even when I added the `trailing: true`, it wouldn't work. And that's because `lodash` [returns the cached results](https://github.com/lodash/lodash/blob/master/debounce.js#L205) on a `debounce` and then [calls the function again](https://github.com/lodash/lodash/blob/master/debounce.js#L150) on a `setTimeout` for the trailing, but because we've already resolved the promise returning the previous results, `loadOptions` on the `Select` is not expecting any more results so the result just seems to get lost 🤷 
I am not 100% sure how this worked before, if you believe we introduced a bug previously I can dig more into that to see how it used to work!

joshhunt commented:
Ahh. That recent PR _undoes_ the change I introduced back in my PR to fix the previous issues. Note the code comment I added:

> when debouncing, we must use the callback form of react-select's loadOptions so we don't drop results for user input. This must not return a promise/use await.

However the refactor changed it back to return a promise, reintroducing the issue (or something similar)

I find it difficult to explain the sequence of events, but I have a poor attempt in the description of my original PR  https://github.com/grafana/grafana/pull/50303. Specifically the third bullet about the trailing call.

joshhunt commented on public/app/features/library-panels/components/OpenLibraryPanelModal/OpenLibraryPanelModal.tsx:
```diff
@@ -31,10 +31,7 @@ export function OpenLibraryPanelModal({ libraryPanel, onDismiss }: OpenLibraryPa
-    () => debounce(loadOptions, 300, { leading: true, trailing: true }),
-    [loadOptions]
-  );
+  const debouncedLoadOptions = useMemo(() => debounce(loadOptions, 300, { leading: true }), [loadOptions]);
```
I _think_ we always want `trailing: true` for search, no? Otherwise won't the last letter the user types not be used for search?

I always find debounce/throttle is difficult to wrap my head around 😓

JoaoSilvaGrafana commented:
> When typing super fast it wasn't quite as instantaneous as I had hoped, but it still does the job and fixes the error.

@samjewell This happens because when you type something, it will immediately make a leading call for the search, but if you type something very quickly like "jewe", it will emit one call for "j" and then wait for 300ms after the last character until it does the call for "jewe" to make sure you are not typing any other characters, I think that is just the nature of how debounce works..

joshhunt reviewed (approved):
👍🏻  On this - I think this is the right approach.

Like Sam said, I think it would be neat to try and get some (unit) tests for this given how this has broken frequently, but don't want to hold up too much if testing the timing for this is too tricky.

JoaoSilvaGrafana commented:
I will try to write some tests, if it turns out to be a nightmare I might quit it because I'm not 100% sure how to test debounces 😅

JoaoSilvaGrafana committed:
Adding a test to FolderPicker

JoaoSilvaGrafana committed:
Merge branch 'main' into joao_silva/folder_search_debounce
